Add PSCI driver

carlocaione opened the pull request:
A bit of rationale behind this PR.

@MrVan submitted in #30676 the PR to enable SMP on AArch64. As I wrote in a review to that PR I do not particularly like the approach of using arch-specific code to interface with the PSCI firmware.

PSCI interface needs a proper driver that uses the DTS to detect the correct parameters, especially the conduit (and currently only that). This is particularly important because the choice of having SMC or HVC as conduit really depends on whether we are running an hypervisor or not in EL2, so this must be something configurable at DTS (board) level.

So here it is a proper PSCI driver with support  for PSCI 0.2 and PSCI 1.0, enabled using the DT and also a tiny test to make sure that everything is working correctly.

carlocaione committed:
drivers: PSCI: Add driver and subsystem

Firmware implementing the PSCI functions described in ARM document
number ARM DEN 0022A ("Power State Coordination Interface System
Software on ARM processors") can be used by Zephyr to initiate various
CPU-centric power operations.

It is needed for virtualization, it is used to coordinate OSes and
hypervisors and it provides the functions used for SMP bring-up such as
CPU_ON and CPU_OFF.

A new PSCI driver is introduced to setup a proper subsystem used to
communicate with the PSCI firmware, implementing the basic operations:
get_version, cpu_on, cpu_off and affinity_info.

The current implementation only supports PSCI 0.2 and PSCI 1.0

The PSCI conduit (SMC or HVC) is setup reading the corresponding
property in the DTS node.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
tests: psci: Introduce PSCI test

Add a simple test to the the PSCI driver

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

sandeepbrcm commented on arch/arm/core/aarch64/smccc-call.S:
```diff
@@ -0,0 +1,25 @@
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <arch/cpu.h>
+
```
Just a note here would help someone unfamiliar to the fact that it implements a spec. 
The SMCCC  excerpt to show what parameter means what would also help :) .  Being greedy here.

carlocaione replied:
you are totally right in being greedy here ;) I'll add some comments.

carlocaione replied:
@sandeepbrcm I added some comments and a link to the spec. Let me know if you want me to add anything else.

MrVan commented:
@carlocaione Thanks for adding psci support, your code looks good to me. Approved.

carlocaione commented:
Can we merge this?

andrewboie commented on tests/arch/arm/arm_psci/src/main.c:
```diff
@@ -0,0 +1,31 @@
+void test_main(void)
+{
+	ztest_test_suite(psci_func,
+		ztest_unit_test(test_psci_func));
```
suggest running this in userspace with ztest_user_unit_test(), granting permission in test_main beforehand

carlocaione replied:
Done in the new revision

tyrantlizard77 reviewed (commented):
Overall looks great but it would be better to have code coverage of the syscalls by running the test function in user mode, and have the test exercise all of the APIs.

carlocaione commented:
> Overall looks great but it would be better to have code coverage of the syscalls by running the test function in user mode, and have the test exercise all of the APIs.

@andrewboie I addressed your points. The only API I couldn't test was `psci_cpu_off` (because that actually shuts the CPU off).
